src/renderers/dom/fiber/ReactDOMFiberComponent.js

```diff
@@ -766,6 +810,36 @@ var ReactDOMFiberComponent = {
 
     assertValidProps(tag, rawProps, getCurrentFiberOwnerName);
 
+    if (__DEV__) {
+      var extraAttributeNames: Set<string> = new Set();
+      var attributes = domElement.attributes;
+      for (var i = 0; i < attributes.length; i++) {
+        // TODO: Do we need to lower case this to get case insensitive matches?
+        var name = attributes[i].name;
+        switch (name) {
+          // Built-in attributes are whitelisted
+          // TODO: Once these are gone from the server renderer, we don't need
+          // this whitelist aynymore.
+          case 'data-reactroot':
+            break;
+          case 'data-reactid':
+            break;
+          case 'data-react-checksum':
+            break;
+          // Controlled attributes are not validated
+          // TODO: Only ignore them on controlled tags.
+          case 'value':
+            break;
+          case 'checked':
+            break;
+          case 'selected':
+            break;
```

Comment:
We fall through elsewhere https://github.com/facebook/react/blob/master/src/renderers/shared/fiber/ReactFiberBeginWork.js#L864-L869